include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -131,8 +137,14 @@ class arena_memory_resource final : public device_memory_resource {
   {
     if (bytes <= 0) return nullptr;
 
-    bytes = detail::arena::align_up(bytes);
-    return get_arena(stream).allocate(bytes);
+    bytes         = detail::arena::align_up(bytes);
+    void* pointer = get_arena(stream).allocate(bytes);
+    if (pointer == nullptr) {
+      if (dump_log_on_failure_) { dump_memory_log(bytes); }
```

Comment:
On my prior comment, I'd add after this line an expensive loop: "for each stream, return all non-superblocks to let the global arena coalesce".

Reply:
It seems we have to be able to achieve this, else the arena allocator will eventually fragment to the point of OOM. Or am I misunderstanding?